### PR TITLE
PCHR-1273: My Leave Calendar Bugfix

### PR DESCRIPTION
**Pull request by deb1990**
This PR contains multiple bug fixes for `my-leave-calendar`

## Bug 1
**Requirement:** The legend should display only the enabled Absence Types
**Comments:** This bug is fixed for both My-Leave and Manager-Leave Calendar.
**Before:**: 
screenshot-3

**After:** 
my_leave___hr17

## Bug 2
**Requirement:** Absence periods in the dropdown should be in chronological order
**Comments:** This bug is fixed for both My-Leave/Manager-Leave Report/Calendar.
**Before:**: 
screenshot-11

**After:** 
my_leave___hr17

## Bug 3
**Requirement:** All the months should be selectable in the months dropdown
**Comments:** This bug is fixed for both My-Leave and Manager-Leave Calendar.
**Before:**: 
screenshot-12

**After:** 
my_leave___hr17

## Bug 4
**Requirement:** Recording late should be hidden as it is not part of the release yet.
**Comments:** This bug is fixed for both My-Leave and Manager-Leave Calendar.
**Before:**: 
screenshot-13

**After:** 
my_leave___hr17

## Bug 5
**Requirement:** Leave requests in calendar should be clickable and should open edit modal
**Screencast:**

## Comments
BackstopJS test cases added for `my-leave-calendar`.

---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/leave-absences/staff/my-leave-calendar.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+'use strict';
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/leave-absences/staff/my-leave-calendar.js]: i believe it was agreed that the folder was `ssp-leave-absences`?

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/leave-absences/staff/my-leave-calendar.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+
+var page = require('../../../page-objects/ssp-leave-absences');
+module.exports = function (casper) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/leave-absences/staff/my-leave-calendar.js]: leave an empty line

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/leave-absences/staff/my-leave-calendar.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+  page.init(casper).initPage()
+    .then(function () {
+      page.init(casper).allMonths();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/leave-absences/staff/my-leave-calendar.js]: it's not clear what this does. All months what? Select all months? Show all months? i checked the method itself and it's not clear also from there

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+var Promise = require('es6-promise').Promise;
+var page = require('./page');
+
+module.exports = (function () {
+
+  return page.extend({
+    /**
+     * [init page]
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: why are you using []?

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+     */
+    initPage: function () {
+      var casper = this.casper;
+
+      return new Promise(function (resolve) {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: I think what you're doing here is another implementation of what is already happening in the main page's object
```js
/**
 * Initializes the page
 *
 * Stores a customized version of CasperJS and then wait for a
 * until a certain "ready" condition is met, if the page is set up to do so
 */
init: function (casper, clearDialogs) {
  !!this.waitForReady && this.casper.then(function () {
    this.waitForReady();
  }.bind(this));
}
```
if that's the case, then this is not necessary. Please check what the current architecture offers before adding new methods

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        casper.then(function () {
+          casper.wait(10000);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: why are you arbitrarily waiting for 10000 seconds (which is a lot)? isn't there something that you can specifically target with a `waitUntilVisibile`?

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+          resolve();
+        });
+      });
+    },
+    /**
+     * [when all months visible]
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: As said before, it's not clear what this method is doing

File: uk.co.compucorp.civicrm.hrcore/backstop_data/scenarios/ssp-la-my-leave-calendar.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
+{
+  "scenarios": [
+    {
+      "label": "Current Month: My Leave Calendar",
+      "url": "my-leave#/my-leave/calendar",
+      "delay": 10000
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/scenarios/ssp-la-my-leave-calendar.json]: This is a lot of delay. Please add a script that waits for a particular element to be visible

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+var Promise = require('es6-promise').Promise;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: the file is called `ssp-leave-absences`, but this is for the my leave page only, so i think the name should be something like `ssp-la-my-leave`?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 10]
```diff
@@ -14,10 +14,10 @@ define([
       return settings.pathTpl + 'components/manager-leave-calendar.html';
     }],
     controllerAs: 'calendar',
-    controller: ['$controller', '$log', '$q', '$rootScope', 'Calendar', 'Contact', 'OptionGroup', 'LeaveRequest', controller]
+    controller: ['$controller', '$log', '$rootScope', '$q', 'Calendar', 'Contact', 'OptionGroup', controller]
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: not sure why you swapped `$rootScope`, `$q`, it should be as it was before

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 11]
```diff
@@ -110,6 +111,8 @@ define([
      * @param  {Array} leaveRequests - leave requests array from API
      */
     vm._indexLeaveRequests = function (leaveRequests) {
+      vm.leaveRequests = {};
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: 1. At the beginning you are initializing `vm.leaveRequests` as an array (`vm.leaveRequests = []`), but now you reset it as an object? Something's not right
2. Why the re-initialization in the first place? It's not clear why it is necessary

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-requests.js [hunk 12]
```diff
@@ -249,7 +249,7 @@ define([
     function loadAbsencePeriods() {
       return AbsencePeriod.all()
         .then(function (absencePeriods) {
-          vm.absencePeriods = absencePeriods;
+          vm.absencePeriods = _.sortBy(absencePeriods, 'title');;
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-requests.js]: You can't sort by the title, a period can have any arbitrary title. You need to use the start date

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js [hunk 13]
```diff
@@ -68,6 +69,8 @@ define([
      * @param  {Array} leaveRequests - leave requests array from API
      */
     vm._indexLeaveRequests = function (leaveRequests) {
+      vm.leaveRequests = {};
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js]: same as before, why is this reset needed?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js [hunk 14]
```diff
@@ -93,7 +96,7 @@ define([
      * @return {Promise}
      */
     vm._loadLeaveRequestsAndCalendar = function () {
-      return parentCtrl._loadLeaveRequestsAndCalendar.call(vm, 'contact_id', true);
+      return parentCtrl._loadLeaveRequestsAndCalendar.call(vm, 'contact_id', false);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js]: isn't there a default value for the second parameter? so that you can omit specifying it?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js [hunk 15]
```diff
@@ -241,7 +241,7 @@ define([
     function loadAbsencePeriods() {
       return AbsencePeriod.all()
         .then(function (absencePeriods) {
-          vm.absencePeriods = absencePeriods;
+          vm.absencePeriods = _.sortBy(absencePeriods, 'title');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-report.js]: same as before

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js [hunk 16]
```diff
@@ -25,8 +25,8 @@ define([
       calendar: false,
       page: false
     };
-    this.monthLabels = ['January', 'February', 'March', 'April', 'May', 'June',
-      'July', 'August', 'September', 'October', 'November', 'December'];
+    this.monthLabels = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]:
```js
this.monthLabels = moment.monthsShort()
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js [hunk 17]
```diff
@@ -231,7 +231,7 @@ define([
     this._loadAbsencePeriods = function () {
       return AbsencePeriod.all()
         .then(function (absencePeriods) {
-          this.absencePeriods = absencePeriods;
+          this.absencePeriods = _.sortBy(absencePeriods, 'title');
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: see previous comments

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js [hunk 18]
```diff
@@ -248,7 +248,9 @@ define([
     this._loadAbsenceTypes = function () {
       return AbsenceType.all()
         .then(function (absenceTypes) {
-          this.absenceTypes = absenceTypes;
+          this.absenceTypes = absenceTypes.filter(function (absenceType) {
+            return absenceType.is_active === "1";
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]:
```js
return absenceType.is_active;
```
should be enough, if you use `===` you might encounter problem in case the api returns 1 as an int

Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: Actually, shouldn't you make the api call with `{ is_active = 1 }`?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js [hunk 19]
```diff
@@ -326,6 +328,15 @@ define([
     };
 
     /**
+     * Reset the months data for before refresh
+     */
+    this._resetMonths = function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: Can you remind me

1. What's inside each `month`
2. When you need to reset whatever it's inside each `month`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js [hunk 20]
```diff
@@ -147,8 +149,17 @@
             expect(controller.levelTypes).toEqual(optionGroupMock.getCollection('hrjc_level_type'));
           });
 
-          it('absence periods have loaded', function () {
-            expect(controller.absencePeriods.length).not.toBe(0);
+          describe('absence periods', function() {
+            it('absence periods have loaded', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js]: always try to makes a sentence with `it`, to improve readibility of the tests
```js
it('loads the absence periods'
```
?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js [hunk 21]
```diff
@@ -147,8 +149,17 @@
...
+              expect(controller.absencePeriods.length).not.toBe(0);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js]: are you using mocked data for the absence periods? if yes, can you have a more meaningful test then "the amount is not zero"? ideally you should test that the controller has the exact amount of periods that you expect it to have

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js [hunk 22]
```diff
@@ -147,8 +149,17 @@
...
+            });
+
+            it('sorts absence periods by title', function () {
+              var extractTitles = function (period) {
+                return period.title;
+              };
+              expect(controller.absencePeriods.map(extractTitles)).toEqual(_.sortBy(absencePeriodData.all().values, 'title').map(extractTitles));
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js]: can you put all that you're doing in `toEqual` in a variable?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/data/absence-period-data.js [hunk 23]
```diff
@@ -15,6 +15,12 @@ define(function () {
       "start_date": "2017-01-01",
       "end_date": "2017-12-31",
       "weight": "2"
+    }, {
+      "id": "3",
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/mocks/data/absence-period-data.js]: why do you need this extra period?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js [hunk 24]
```diff
@@ -82,8 +83,14 @@
...
 
-        it('absence periods have loaded', function () {
-          expect(controller.absencePeriods.length).not.toBe(0);
+        describe('absence periods', function() {
+          it('absence periods have loaded', function () {
+            expect(controller.absencePeriods.length).not.toBe(0);
+          });
+
+          it('sorts absence periods by title', function () {
+            expect(controller.absencePeriods).toEqual(_.sortBy(absencePeriodData.all().values, 'title'));
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js]: why is the `toEqual` argument less complex that the one that you're using in the manager leave request test?

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/manager-leave/components/manager-leave-calendar.html [hunk 25]
```diff
@@ -185,14 +184,15 @@ <h2 class="panel-title" role="button" ng-click="calendar.legendCollapsed = !cale
...
             </div>
-            <div class="col-sm-3">
-              <div class="chr_leave-calendar__legend__entry">
-                <div
-                  class="chr_leave-calendar__item chr_leave-calendar__item--type--late chr_leave-calendar__item--alt">
-                  <span>L</span></div>
-                <span class="chr_leave-calendar__legend__label">Late</span>
-              </div>
-            </div>
+            <!--Uncomment when LATE is developed -->
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/manager-leave/components/manager-leave-calendar.html]: Remove this, if we need it we can take it from the commit history

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-calendar-month.html [hunk 26]
```diff
@@ -30,6 +30,12 @@ <h2 class="panel-title">{{calendar.monthLabels[month.month] + " " + month.year}}
             <span ng-if="day.UI.isAM">AM</span>
             <span ng-if="day.UI.isPM">PM</span>
             <span ng-if="day.UI.isAccruedTOIL">AT</span>
+            <a ng-if="day.leaveRequest"
+               contact-id="calendar.contactId"
+               leave-request-popup
+               leave-request="day.leaveRequest"
+               title="Click to open Leave Request">
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-calendar-month.html]: "Click to open this leave request"

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-calendar.html [hunk 27]
```diff
@@ -133,14 +132,15 @@ <h2 class="panel-title" role="button" ng-click="calendar.legendCollapsed = !cale
...
             </div>
-            <div class="col-sm-3">
-              <div class="chr_leave-calendar__legend__entry">
-                <div
-                  class="chr_leave-calendar__item chr_leave-calendar__item--type--late chr_leave-calendar__item--alt">
-                  <span>L</span></div>
-                <span class="chr_leave-calendar__legend__label">Late</span>
-              </div>
-            </div>
+            <!--Uncomment when LATE is developed -->
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-calendar.html]: Same as before

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 11 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: 1. Fixed
2. It is necessary because when the calendar reloads because of a leave request is created/edited `vm.leaveRequests` should clear before reassigning values, otherwise duplicate values are shown in the UI.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js (hunk 13 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js]: It is necessary because when the calendar reloads because of a leave request is created/edited vm.leaveRequests should clear before reassigning values, otherwise duplicate values are shown in the UI.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js (hunk 14 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/my-leave/components/my-leave-calendar.js]: Default is true, so I need pass `false`.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js (hunk 18 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: Implemented the filter in the API level

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js (hunk 19 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: 1. We have a structure like this
```
{
  month: startDate.month(),
  year: startDate.year(),
  data: []
}
```

2. It is necessary because when the calendar reloads because of a leave request is created/edited the `data` should clear before reassigning values, otherwise duplicate values are shown in the UI.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js (hunk 24 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js]: For manager-leave-request a key called `current` is getting added which is preventing me from directly equaling the values. So extracting the titles.

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js (hunk 5 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: No required anymore

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js (hunk 4 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-leave-absences.js]: Other files had [], and I followed that. Later found that that [] was present as la-milestone-3 is not synced with 1.6. Removed the [].

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+     */
+    showAllMonths: function () {
+      var casper = this.casper;
+      //If Current month is december then need to wait for November as it would be loaded last when all months load
+      var lastMonth = currentMonth === 12 ? 11 : 12;
+
+      casper.then(function () {
+        //Wait for the current month spinner to be first visible then hidden
+        casper.waitUntilVisible('.panel.panel-default:nth-of-type(' + currentMonth + ') ng-include .chr_spinner', function () {
+          casper.waitWhileVisible('.panel.panel-default:nth-of-type(' + currentMonth + ') ng-include .chr_spinner', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js]: if you put `waitForCurrentMonthsVisibility` in `waitForReady` as in previous comments, the page will automatically use it everytime you `init` it, so you don't have to repeat this in every method

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js [hunk 29]
```diff
@@ -0,0 +1,44 @@
...
+            casper.click('.panel-heading .close.ui-select-match-close');
+            //Wait for the last month spinner to be first visible then hidden
+            casper.waitUntilVisible('.panel.panel-default:nth-of-type(' + lastMonth + ') ng-include', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js]: if this "wait for it to be visible, then to be hidden" logic cannot be avoided, then please put it in a private function, cause you have lots of repetition with it

File: uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/ssp-leave-absences/staff/my-leave-calendar_current-month-visible.js [hunk 30]
```diff
@@ -0,0 +1,7 @@
+'use strict';
+
+var page = require('../../../page-objects/ssp-la-my-leave');
+
+module.exports = function (casper) {
+  page.init(casper).waitForCurrentMonthsVisibility();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/casper_scripts/ssp-leave-absences/staff/my-leave-calendar_current-month-visible.js]: if you use the `waitForReady` method (see other comment), this can simply be
```js
page.init(casper);
```
because `init` will automatically call `waitForReady`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js [hunk 31]
```diff
@@ -111,7 +111,7 @@ define([
      * @param  {Array} leaveRequests - leave requests array from API
      */
     vm._indexLeaveRequests = function (leaveRequests) {
-      vm.leaveRequests = {};
+      vm.leaveRequests = [];
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: This is more confusing to be honest, i was sure that `vm.leaveRequest = []` was wrong. You are doing
```js
vm.leaveRequests[leaveRequest.contact_id] = // something
```
how can `vm.leaveRequest` be an array?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js [hunk 32]
```diff
@@ -25,8 +25,7 @@ define([
       calendar: false,
       page: false
     };
-    this.monthLabels = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun',
-      'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
+    this.monthLabels = moment.monthsShort();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: this is now a one-liner, put it together with the others
```js
this.months = [];
this.monthLabel = moment.monthsShort();
this.selectedMonths = [];
// etc
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js [hunk 33]
```diff
@@ -246,12 +245,11 @@ define([
...
      */
     this._loadAbsenceTypes = function () {
-      return AbsenceType.all()
-        .then(function (absenceTypes) {
-          this.absenceTypes = absenceTypes.filter(function (absenceType) {
-            return absenceType.is_active === "1";
-          });
-        }.bind(this));
+      return AbsenceType.all({
+        is_active: true
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/calendar-ctrl.js]: Update the method description

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js [hunk 34]
```diff
@@ -150,15 +150,16 @@
...
-            it('sorts absence periods by title', function () {
-              var extractTitles = function (period) {
-                return period.title;
+            it('sorts absence periods by start_date', function () {
+              var extractStartDate = function (period) {
+                return period.start_date;
               };
-              expect(controller.absencePeriods.map(extractTitles)).toEqual(_.sortBy(absencePeriodData.all().values, 'title').map(extractTitles));
+              var absencePeriodSortedByDate = _.sortBy(absencePeriodData.all().values, 'start_date').map(extractStartDate);
+              expect(controller.absencePeriods.map(extractStartDate)).toEqual(absencePeriodSortedByDate);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/manager-leave/components/manager-leave-requests_test.js]: add an empty line between var declarations and rest of the code

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js [hunk 35]
```diff
@@ -67,6 +68,7 @@
 
           return $q.resolve(data);
         });
+        spyOn(AbsenceType, 'all').and.callThrough();
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js]: i don't think you need to spy on `AbsenceType` for every test, right?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js [hunk 36]
```diff
@@ -118,11 +118,12 @@
...
-              it('sorts absence periods by title', function () {
-                var extractTitles = function (period) {
-                  return period.title;
+              it('sorts absence periods by start_date', function () {
+                var extractStartDate = function (period) {
+                  return period.start_date;
                 };
-                expect(controller.absencePeriods.map(extractTitles)).toEqual(_.sortBy(absencePeriodData.all().values, 'title').map(extractTitles));
+                var absencePeriodSortedByDate = _.sortBy(absencePeriodData.all().values, 'start_date').map(extractStartDate);
+                expect(controller.absencePeriods.map(extractStartDate)).toEqual(absencePeriodSortedByDate);
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-report_test.js]: add empty line

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-calendar-month.html [hunk 37]
```diff
@@ -34,7 +34,7 @@ <h2 class="panel-title">{{calendar.monthLabels[month.month] + " " + month.year}}
                contact-id="calendar.contactId"
                leave-request-popup
                leave-request="day.leaveRequest"
-               title="Click to open Leave Request">
+               title="Click to open this Leave Request">
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/my-leave/components/my-leave-calendar-month.html]: "leave request", it's not a proper noun, so it shouldn't be capitalized

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+  return page.extend({
+    /**
+     * Wait for current month to be visible
+     */
+    waitForCurrentMonthsVisibility: function () {
+      var casper = this.casper;
+
+      casper.then(function () {
+        //Wait for the current month spinner to be first visible then hidden
+        casper.waitUntilVisible('.panel.panel-default:nth-of-type(' + currentMonth + ') ng-include .chr_spinner', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js]: also remember to leave a space after `//`
```js
// Wait for
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js (hunk 35 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/my-leave/components/my-leave-calendar_test.js]: True, but as `AbsenceType.all`gets called while init. I had to put it before component compilation

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js (hunk 31 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/manager-leave/components/manager-leave-calendar.js]: Yes it should be object, as both are same, moved to parent controller.

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js (hunk 38 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js]: Not required anymore

File: uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js (hunk 29 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrcore/backstop_data/page-objects/ssp-la-my-leave.js]: Not required anymore
